util: make the qsort_r check work on macOS

tiennou opened the pull request:
As it seems, there are 2 versions of `qsort_r` (a non-standard function) out there : the BSD version, and the glibc version, and they differ in their parameter ordering (which is quite dangerous because using one for the other will result in runtime failures).

Right now, the `defined(BSD)` check makes us use the "default" version on macOS, even though it has a BSD implementation available, so the check needs to be weakened. An alternative would be to perform a "BSD-style" compilation check in CMake, and act accordingly.

tiennou commented:
I found it while looking at the tsort implementation, which has worried Coverity for a loooong time (which I think it's actually safe and Coverity is just paranoid). As a heads-up, I've also closed as Intentional some of the defects there (around git_buf_oom usage, and some of the ones I remember taking a hard look at).

tiennou commented on src/util.c:
```diff
@@ -647,7 +647,7 @@ size_t git__unescape(char *str)
 }
 
-#if defined(HAVE_QSORT_S) || (defined(HAVE_QSORT_R) && defined(BSD))
+#if defined(HAVE_QSORT_S) || (defined(HAVE_QSORT_R) && !defined(__GLIBC__))
```
TL;DR every platform is weird 😉. I think it's doable to make a `CHECK_C_COMPILE` test in CMake that would detect the parameter ordering, which would make it at least C-library-resistant (that's the alternative I proposed).

AFAICT, here is what I've been able to gather :
- qsort_r only exists in [FreeBSD/DragonFly](http://bxr.su/search?q=qsort_r&defs=&refs=&path=&project=DragonFly&project=FreeBSD&project=NetBSD&project=OpenBSD), ie. it's not *really* `defined(BSD)`, and percolated through to macOS.
- Windows has a `qsort_s` version.
- This was helpful [comment](https://github.com/kripken/emscripten/pull/5424#issuecomment-318511870), as well as [this one](https://stackoverflow.com/a/39561369/444323).

I'll try to see how doable it is to detect all those qsort-ses using compilation checks.

tiennou commented:
Rebased, checking the prototype is done at compile-time now.

pks-t commented on src/CMakeLists.txt:
```diff
@@ -61,6 +61,29 @@ ENDIF ()
+" HAVE_QSORT_R_BSD)
+	IF (HAVE_QSORT_R_BSD)
+		ADD_DEFINITIONS(-DHAVE_QSORT_R_BSD)
+	ENDIF ()
```
I think this approach is a lot more sensible. than what we currently have. I'd like to propose a few changes, though:

- instead of defining `HAVE_QSORT_R`, only ever define `HAVE_QSORT_R_BSD` and `HAVE_QSORT_R_GNU` (and adjust locations using the previous define accordingly)
- instead of `CHECK_C_SOURCE_COMPILES` and/or `CHECK_FUNCTION_EXISTS`, we should be using `CHECK_PROTOTYPE_DEFINITION`

So, something like this:

```
CHECK_PROTOTYPE_DEFINITION(qsort_r
	"void qsort_r(void *base, size_t nmemb, size_t size, void *thunk, int (*compar)(void *, const void *, const void *))"
	"NULL" "stdlib.h" HAVE_QSORT_R_BSD)
CHECK_PROTOTYPE_DEFINITION(qsort_r
	"void qsort_r(void *base, size_t nmemb, size_t size, int (*compar)(const void *, const void *, void *), void *arg)"
	"NULL" "stdlib.h" HAVE_QSORT_R_GNU)
```

The only thing is that we'd have to include `CheckPrototypeDefinition`, which is only available since CMake v3.0. There's nothing speaking against distributing it in our cmake/Modules directory, though.

tiennou commented:
Rebased, and changed to use the CMake module, which I lifted from [this](https://github.com/Kitware/CMake/commit/4f252abea5f1d17c60f6ff115c9c44cc0b6f1df6). I've tested  2.8.11, and it's okay with configuring that, but it didn't build though :

```
clang: error: no such file or directory: '/Users/tiennou/Projects/libgit2/_build/deps/http-parser/libgit2.build/Debug/http-parser.build/Objects-normal/undefined_arch/http_parser.o'
```

`undefined_arch` makes me think that it doesn't support Xcode 9 😉.
